runtime/colors/vim.lua

```diff
@@ -0,0 +1,290 @@
+-- Name:         vim
+-- Description:  Vim's default color scheme
+-- Author:       Bram Moolenaar <Bram@vim.org>
+-- Maintainer:   The Vim Project <https://github.com/vim/vim>
+-- Website:      https://github.com/vim/vim
+-- License:      Same as Vim
+-- Last Updated: 2023 Aug 10
+
+-- This is Vim's default color scheme. It doesn't define the Normal
+-- highlighting, it uses whatever the colors used to be.
+
+vim.cmd('hi clear')
+vim.g.colors_name = 'vim'
+
+local set_hl, cmd = vim.api.nvim_set_hl, vim.cmd
+local hi = function(name, val)
+  -- Force links
+  val.force = true
+
+  -- Make sure that `cterm` attribute is not populated from `gui`
+  val.cterm = val.cterm or {}
+
+  -- Define global highlight
+  set_hl(0, name, val)
+end
+
+--stylua: ignore start
+-- General
+hi('Normal', {})
+
+cmd('hi Cursor  guifg=bg guibg=fg')
```

Comment:
This is working around `nvim_set_hl` erroring if `bg` or `fg` are not defined for the normal group? 

What's the rationale for setting these (non-defined) attributes here?

Reply:
`hi('Normal', {})` is mimicking `hi Normal gui=NONE cterm=NONE` from Vimscript version.

Usage of `cmd()` is to work around the fact that 'fg' and 'bg' can't be set as attribute value in `vim.api.nvim_set_hl()`, yes. Otherwise, there are errors.

> What's the rationale for setting these (non-defined) attributes here?

Not sure which attributes is this about?

Reply:
> Not sure which attributes is this about?

`guifg=fg` instead of the equivalent (and arguably more explicit) `guifg=NONE` (or just omitting it, since `nvim_set_hl` clears omitted attributes anyway)

Reply:
> > Not sure which attributes is this about?
> 
> `guifg=fg` instead of the equivalent (and arguably more explicit) `guifg=NONE` (or just omitting it, since `nvim_set_hl` clears omitted attributes anyway)

It is `guifg=bg` and this is the way it was done in Vim's color scheme. So just mimicking here because I don't completely know why it is defined the way it is.

Reply:
> So just mimicking here because I don't completely know why it is defined the way it is.

"20 year old tricky ways", if I had to guess.

I'd _really_ prefer if we could drop the hackery for our files, as long as the outcome is still the same.

Reply:
Me too. The point here is that I can not vouch that using `hi('Cursor', {})` and `vim.cmd.highlight('Cursor  guifg=bg guibg=fg')` result in the same outcome.

I can replace it, though, of course. If all tests still pass, will this be enough?

Reply:
> I can replace it, though, of course. If all tests still pass, will this be enough?

Yes, that would be fine. If someone complains later, we can still revisit (and possibly fix the API if the use case is strong enough).

Reply:
Judging by resolved #25474, having `vim.cmd.highlight('Cursor  guifg=bg guibg=fg')` (or in `Ignore`, for that matter) is different from `vim.cmd.highlight('Cursor  guifg=NONE guibg=NONE')` in case of cleared `Normal`. Mostly because that `"fg"` and `"bg"` colors will be updated after `Normal` is updated.

So I don't see a way of getting rid of those `vim.cmd.highlight()` calls for groups which use `"fg"` or `"bg"` as colors, as `vim.api.nvim_set_hl()` errors in that case.

Reply:
@bfredl is this something we can fix? I think the goal is to have nvim_set_hl be a full alternative to :highlight. 

(Otherwise I am fine with declaring this a &bg=&bg level hackery that was never explicitly promised to work.)

Reply:
> By the way, this is contrary to [what is written in the help](https://github.com/neovim/neovim/blob/49efdf84139a0c0f7ba0217e5ab665bc4f222206/runtime/doc/syntax.txt#L5076-L5078): it seems that changing Normal should not result into "fg" and "bg" colors being changed.

This means that we don't have to support this use case ("live" values of `fg` and `bg`) and `guifg=NONE` replicates the _documented_ behavior of `fg=bg`. So we _can_ (and should) get rid of the `vim.cmd.highlight` calls (except `clear`).

Reply:
Not supporting this case will mean that `Cursor`, `lCursor`, and `Ignore` will not be the same as they are in Vim. I'd argue that this has higher priority than not using `vim.cmd.highlight()`.

The best course of action is to make `nvim_set_hl()` behave the same way as `:hi` in this case.

Reply:
> We don't need to match Vim _exactly_; we just want, e.g., `:colorscheme quiet` (and tests) behave as reasonably expected.

Sure, if you say so.